precompiles/outposts/osmosis/types.go

```diff
@@ -0,0 +1,219 @@
+	// NextMemo contains any additional memo information for the next operation in a PFM setting.
+	NextMemo string `json:"next_memo,omitempty"`
+}
+
+// Msg contains the OsmosisSwap details used in the memo relayed to the Osmosis ibc hook middleware.
+type Msg struct {
+	// OsmosisSwap provides details for a swap transaction. It's optional and can be omitted if not provided.
+	OsmosisSwap *OsmosisSwap `json:"osmosis_swap"`
+}
+
+// Memo wraps the message details for the IBC packet relyaed to the Osmosis chain. This include the
+// address of the smart contract that will receive the Msg.
+type Memo struct {
+	// Contract represents the address or identifier of the contract to be called.
+	Contract string `json:"contract"`
+	// Msg contains the details of the operation to be executed on the contract.
+	Msg *Msg `json:"msg"`
+}
+
+// RawPacketMetadata is the raw packet metadata used to construct a JSON string.
+type RawPacketMetadata struct {
+	// The Osmosis outpost IBC memo.
+	Memo *Memo `json:"memo"`
+}
+
+// CreatePacketWithMemo creates the IBC packet with the memo for the Osmosis
+// outpost that can be parsed by the ibc hook middleware on the Osmosis chain.
+func CreatePacketWithMemo(
+	outputDenom, receiver, contract string,
+	slippagePercentage uint8,
+	windowSeconds uint64,
+	onFailedDelivery string,
+) *RawPacketMetadata {
+	return &RawPacketMetadata{
+		&Memo{
+			Contract: contract,
+			Msg: &Msg{
+				OsmosisSwap: &OsmosisSwap{
+					OutputDenom: outputDenom,
+					Slippage: &Slippage{
+						&Twap{
+							SlippagePercentage: slippagePercentage,
+							WindowSeconds:      windowSeconds,
+						},
+					},
+					Receiver:         receiver,
+					OnFailedDelivery: onFailedDelivery,
+					// NextMemo:         "",
+				},
+			},
+		},
+	}
+}
+
+// ConvertToJSON convert the RawPacketMetadata type into a JSON formatted
+// string.
+func (r RawPacketMetadata) ConvertToJSONString() (string, error) {
+	// Convert the struct to a JSON string
+	jsonBytes, err := json.MarshalIndent(r, "", "  ")
+	if err != nil {
+		log.Fatalf("Failed to marshal JSON: %v", err)
+	}
+
+	return string(jsonBytes), nil
+}
+
+// ValidateSwapToken validates the input and outpost tokens for the swap.
+func ValidateSwapTokens(input, output, stakingDenom, portID, channelID string) error {
+	if input == output {
+		return fmt.Errorf(ErrInputEqualOutput, input)
+	}
+
+	osmoIBCDenom := transfertypes.DenomTrace{
+		Path:      fmt.Sprintf("%s/%s", portID, channelID),
+		BaseDenom: OsmosisDenom,
+	}.IBCDenom()
+
+	// Check that the input token is evmos or osmo.
+	// This constraint will be removed in future
+	validInput := []string{stakingDenom, osmoIBCDenom}
+	if !slices.Contains(validInput, input) {
+		return fmt.Errorf(ErrInputTokenNotSupported, validInput)
+	}
+
+	return nil
+}
+
+// ValidateSwapParameters validates the parameters used to perform the swap.
+func ValidateSwapParameters(slippagePercentage uint8, windowSeconds uint64) error {
```

Comment:
we can convert them into a single function that calls both

```go
func (m RawPacketMetadata) Validate() error
```

Reply:
Concerning this one, the idea was to have small chunks of independent validations that we can easily remove without affecting the single test logic and the single function. In this way we have small functions with relatively few inputs instead of a monolithic one. WDYT?

Reply:
all the types have a single validation function (`ValidateBasic` or `Validate`). We should keep it consistent with that

Reply:
This is true, but they are implemented as methods because are tight to the type itself. In this case this is a function. Do you think that makes sense to add a Validate method to the RawPacketMetadata type with all this validations instead of the function? In this way we maintain consistency as you mentioned.

Reply:
>  Do you think that makes sense to add a Validate method to the RawPacketMetadata type with all this validations instead of the function

yes, that's what I'm proposing